logstash/check.py

```diff
@@ -0,0 +1,319 @@
+                    message="Error {0} when hitting {1}".format(e, url),
+                    tags=config.service_check_tags
+                )
+            raise
+
+        return resp.json()
+
+    def _get_logstash_version(self, config):
+        """ Get the running version of logstash.
+        """
+        try:
+            data = self._get_data(config.url, config, send_sc=False)
+            version = data['version']
+        except Exception as e:
+            self.warning(
+                "Error while trying to get Logstash version "
+                "from %s %s. Defaulting to version 1.0.0."
+                % (config.url, str(e))
+            )
+            version = "1.0.0"
+
+
+        self.service_metadata('version', version)
+        self.log.debug("Logstash version is %s" % version)
+        return version
+
+    def check(self, instance):
+        config = self.get_instance_config(instance)
+
+        version = self._get_logstash_version(config)
+
+        stats_metrics = dict(self.STATS_METRICS)
+        if version and LooseVersion(version) < LooseVersion("6.0.0"):
+            stats_metrics.update(self.PIPELINE_METRICS)
+
+        stats_url = urlparse.urljoin(config.url, '/_node/stats')
+        stats_data = self._get_data(stats_url, config)
+
+        self._process_stats_data(stats_data, stats_metrics, config)
+
+        if version and LooseVersion(version) < LooseVersion("6.0.0"):
+            self._process_pipeline_plugins_data(stats_data['pipeline']['plugins'], self.PIPELINE_INPUTS_METRICS, config, 'inputs', 'input_name')
+            self._process_pipeline_plugins_data(stats_data['pipeline']['plugins'], self.PIPELINE_OUTPUTS_METRICS, config, 'outputs', 'output_name')
+            self._process_pipeline_plugins_data(stats_data['pipeline']['plugins'], self.PIPELINE_FILTERS_METRICS, config, 'filters', 'filter_name')
+
+        self.service_check(
+            self.SERVICE_CHECK_CONNECT_NAME,
+            AgentCheck.OK,
+            tags=config.service_check_tags
+        )
+
+    def _process_stats_data(self, data, stats_metrics, config):
+        for metric, desc in stats_metrics.iteritems():
+            self._process_metric(data, metric, *desc, tags=config.tags)
+
+
+    def _process_pipeline_plugins_data(self, pipeline_plugins_data, pipeline_plugins_metrics, config, plugin_type, tag_name):
+        for plugin_data in pipeline_plugins_data.get(plugin_type,[]):
+            plugin_name = plugin_data.get('name')
+
+            metrics_tags = list(config.tags)
+
+            if not plugin_name:
+                    plugin_name = 'unknown'
+
+            metrics_tags.append(
+                u"{}:{}".format(tag_name, plugin_name)
+            )
+
+            for metric, desc in pipeline_plugins_metrics.iteritems():
+                self._process_metric(
+                    plugin_data, metric, *desc,
+                    tags=metrics_tags)
+
+    def _process_pipeline_input_data(self, data, pipeline_inputs_metrics, config):
```

Comment:
since we're using `_process_pipeline_plugins_data` we should remove `_process_pipeline_input_data`, `_process_pipeline_output_data` and `_process_pipeline_filters_data`.